providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,334 @@
+#!/usr/bin/env python3
+"""prime offload test module."""
+# This file is part of Checkbox.
+#
+# Copyright 2023 Canonical Ltd.
+# Written by:
+#   Hanhsuan Lee <hanhsuan.lee@canonical.com>
+#
+# Checkbox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# Checkbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import threading
+import subprocess
+import time
+import re
+import json
+import argparse
+import logging
+from enum import IntEnum
+
+
+class PrimeOffloaderError(IntEnum):
+    """
+    A class used to define PrimeOffloader Error code
+
+    :attr NO_ERROR: process success
+    :type NO_ERROR: int
+
+    :attr NO_CARD_ID: couldn't find card id
+    :type NO_CARD_ID: int
+
+    :attr NO_CARD_NAME: couldn't find card name
+    :type NO_CARD_NAME: int
+
+    :attr OFFLOAD_FAIL: couldn't find process on specific GPU
+    :type OFFLOAD_FAIL: int
+
+    :attr NOT_SUPPORT_NV_PRIME: this system couldn't support nv prime offload
+    :type NOT_SUPPORT_NV_PRIME: int
+
+    :attr NV_DRIVER_ERROR: some errors form nvidia driver
+    :type NV_DRIVER_ERROR: int
+
+    """
+
+    NO_ERROR = 0
+    NO_CARD_ID = -1
+    NO_CARD_NAME = -2
+    OFFLOAD_FAIL = -3
+    NOT_SUPPORT_NV_PRIME = -4
+    NV_DRIVER_ERROR = -5
+
+
+class PrimeOffloader:
+    """
+    A class used to execute process to specific GPU.
+    Have to run this as root.
+
+    :attr logger: console logger
+    :type logger: obj
+
+    :attr check_result:
+        store the result of checking offloading is ok or not.
+    :type check_result: int
+    """
+
+    logger = logging.getLogger()
+
+    check_result = PrimeOffloaderError.OFFLOAD_FAIL
+
+    def find_card_id(self, pci_name):
+        """
+        use pci name to find card id under /sys/kernel/debug/dri
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+        """
+        cmd = "sudo grep -lr --include=name \"{}\""  \
```

Comment:
1. Please don't escape line endings. Use inline string concatenations instead.
2. The preferred way of calling subprocesses is to use a list of arguments, and considering this cmd is getting interpolated with other strings, it will be unsafe to run with `shell=True` as there is no lexical check.
3. If the program is meant to be run as root, don't invoke sudo, as it leads to unpredictable results.